crates/datasources/src/sqlserver/client.rs

```diff
@@ -0,0 +1,209 @@
+            match req {
+                Request::Query { query, response } => {
+                    let mut stream = match self.client.query(query, &[]).await {
+                        Ok(stream) => stream,
+                        Err(e) => {
+                            // We don't care if this errors, just means that the
+                            // client is no longer listening.
+                            let _ = response.send(Err(e.into())).await;
+                            continue;
+                        }
+                    };
+
+                    while let Some(item) = stream.next().await {
+                        if response.send(item.map_err(|e| e.into())).await.is_err() {
+                            // Client no longer listening, stop reading from the
+                            // stream and just move to the next message.
+                            break;
+                        }
+                    }
+                }
+                Request::Drop => {
+                    debug!("closing SQL Server connection");
+                    self.client.close().await?;
+                    return Ok(());
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Client side of the connection.
+#[derive(Debug)]
+pub struct Client {
+    sender: mpsc::UnboundedSender<Request>,
+}
+
+impl Client {
+    fn new(sender: mpsc::UnboundedSender<Request>) -> Self {
+        Client { sender }
+    }
+
+    // TODO: Zero clue how we'd want to handle params. Might as well just fork
+    // tiberius to make it more reasonable.
+    pub async fn query<'a>(&self, query: impl Into<Cow<'a, str>>) -> Result<QueryStream> {
+        let query = query.into().to_string();
+
+        let (sender, receiver) = mpsc::channel(1);
+        let req = Request::Query {
+            query,
+            response: sender,
+        };
+
+        if self.sender.send(req).is_err() {
+            return Err(SqlServerError::String(
+                "connection to SQL Server closed".to_string(),
+            ));
+        }
+
+        Ok(QueryStream {
+            receiver,
+            metadata: None,
+            buffered_rows: VecDeque::new(),
+        })
+    }
+}
+
+impl Drop for Client {
+    fn drop(&mut self) {
+        let _ = self.sender.send(Request::Drop);
+    }
+}
+
+#[derive(Debug)]
+enum Request {
+    /// Run a query, sending response items to the channel.
+    Query {
+        query: String,
+        response: mpsc::Sender<Result<QueryItem>>,
+    },
+    /// Client was dropped, drop the connection.
+    Drop,
+}
+
+/// Streaming result of a query.
+///
+/// Note that the stream implementation for this returns _only_ rows. The
+/// `QueryStream` in tiberius will return `QueryItem`, however that's not useful
+/// in our case, we just care about the rows.
+#[derive(Debug)]
+pub struct QueryStream {
+    receiver: mpsc::Receiver<Result<QueryItem>>,
+    metadata: Option<ResultMetadata>,
+    buffered_rows: VecDeque<Row>,
```

Comment:
We're unbounded in the mpsc from `Client` to `Connection`, but bounded in the mpsc from `Connection` to `QueryStream` (the results). `buffered_rows` is unbounded, but is effectively 0 (no allocations) or 1 (unlikely).

For the unbounded mpsc, a more proper implementation would send messages directly to the server as soon as it's received instead of after reading the entirety of the response for a message before moving to the next one. The current implementation is still correct, but would unexpectedly block on 2 client messages. We're only sending 1 message at a time anyways so that's not really an issue for us right now though.